### PR TITLE
Core: Replace `fs-extra` with the native APIs

### LINKED ISSUE #29104: Migrate from `fs-extra` to `node:fs`
Originally proposed by @lukeed on Discord.

> You don’t need fs-extra at all
> It 99% implements APIs that are now just node natives
> It’s a legacy package thru and thru

### PR DESCRIPTION
**Pull request by pralkarz**
Closes #29104. Works on #29038.

## What I did

This PR replaces `fs-extra` with the native APIs in all `code/**/*` packages. Now it's only included in `scripts/` and as a transitive dependency.

### Testing

I've confirmed that the unit tests pass locally in the packages I've made the changes to. Please advise on which integration/E2E tests I should run (if any).

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [X] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

Here's what I did for most changes (in packages that have had the unit tests):
1. Run `npx vitest` in the package directory before making any changes.
2. Build in watch mode.
3. Make changes and either rerun `npx vitest` after every change (for smaller packages) or utilize Vitest's watching capabilities (for bigger packages).

As above, please advise if I should do some more testing here.

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  77.4 MB | 77.4 MB | 0 B | **2.22** | 0% |
| initSize |  163 MB | 162 MB | -507 kB | -0.23 | -0.3% |
| diffSize |  85.4 MB | 84.9 MB | -507 kB | -0.27 | -0.6% |
| buildSize |  7.57 MB | 7.57 MB | -2 B | 0.8 | 0% |
| buildSbAddonsSize |  1.66 MB | 1.66 MB | 0 B | 0.89 | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  2.34 MB | 2.34 MB | 0 B | 0.82 | 0% |
| buildSbPreviewSize |  352 kB | 352 kB | 0 B | 0.65 | 0% |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  4.55 MB | 4.55 MB | 0 B | 0.89 | 0% |
| buildPreviewSize |  3.02 MB | 3.02 MB | -2 B | 0.11 | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  25.4s | 13s | -12s -412ms | 0.03 | -95.3% |
| generateTime |  19.7s | 20.6s | 900ms | -0.13 | 4.4% |
| initTime |  16.1s | 18.8s | 2.7s | **1.33** | 🔺14.7% |
| buildTime |  10.4s | 10.5s | 146ms | -0.88 | 1.4% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  7.9s | 7.7s | -277ms | 0.82 | -3.6% |
| devManagerResponsive |  5.1s | 5.2s | 18ms | 1.23 | 0.3% |
| devManagerHeaderVisible |  874ms | 771ms | -103ms | -0.37 | -13.4% |
| devManagerIndexVisible |  915ms | 802ms | -113ms | -0.41 | -14.1% |
| devStoryVisibleUncached |  1.3s | 1.3s | 26ms | -0.09 | 1.9% |
| devStoryVisible |  914ms | 801ms | -113ms | -0.42 | -14.1% |
| devAutodocsVisible |  776ms | 671ms | -105ms | -0.55 | -15.6% |
| devMDXVisible |  778ms | 679ms | -99ms | -0.43 | -14.6% |
| buildManagerHeaderVisible |  904ms | 720ms | -184ms | -0.57 | -25.6% |
| buildManagerIndexVisible |  944ms | 721ms | -223ms | -0.74 | -30.9% |
| buildStoryVisible |  945ms | 834ms | -111ms | 0.03 | -13.3% |
| buildAutodocsVisible |  898ms | 676ms | -222ms | -0.42 | -32.8% |
| buildMDXVisible |  726ms | 679ms | -47ms | -0.06 | -6.9% |

## Greptile Summary

This PR replaces the `fs-extra` package with native Node.js `fs` and `fs/promises` APIs across multiple files in the Storybook codebase, aiming to reduce dependencies and use built-in functionality.

- Replaced `fs-extra` with `node:fs` and `node:fs/promises` in core packages and builders
- Updated file operations to use native methods like `readFile`, `writeFile`, `cp`, and `rm`
- Removed `fs-extra` from package.json files in affected packages
- Added TODO comments regarding experimental `fsPromises.cp` usage in Node 16-21
- Implemented custom file existence checks to replace `fs-extra`'s `ensureFile` functionality

### CODE REVIEW

File: code/builders/builder-vite/src/index.ts
Comment (ziebam) [inline code/builders/builder-vite/src/index.ts:88]: You'll notice this comment in several places in this PR. Since `cp` and `cpSync` are not experimental only from Node 22 onwards, should we:
1. Use them anyway?
2. Use `fs-extra` in these cases?
3. Use a combination of `fs.stat`, `fs.mkdir`, and `fs.open` instead of `cp`/`cpSync`?

File: code/__mocks__/fs/promises.ts
Comment (ziebam) [reply code/__mocks__/fs/promises.ts:16]: False positive.

File: code/builders/builder-vite/src/index.ts
Comment (43081j) [reply code/builders/builder-vite/src/index.ts:88]: From what I understand, it was marked as experimental but did exist for a while in node versions before 22 right? 

i.e earlier versions do support it whether it is considered experimental or not

If that's true, I wonder if we should just use it anyway

Let's wait for someone else to review too though

File: code/core/scripts/helpers/dependencies.ts
Comment (43081j) [reply code/core/scripts/helpers/dependencies.ts:21]: I think this is right and we need to update to be `utf8`

File: code/core/src/cli/helpers.ts
Comment (43081j) [inline code/core/src/cli/helpers.ts:202]: Should we keep this async maybe? Assuming there's an async fs method for checking existence

File: code/core/scripts/helpers/dependencies.ts
Comment (ziebam) [reply code/core/scripts/helpers/dependencies.ts:21]: There are other usages of `encoding: 'utf-8'` in the repository, so it seems like that works too, but I agree that we should be consistent with the documentation. Going to change it.

File: code/core/src/cli/helpers.ts
Comment (ziebam) [reply code/core/src/cli/helpers.ts:202]: There's no Promise-based alternative, and the callback one is marked as deprecated: https://nodejs.org/docs/latest-v18.x/api/fs.html#fsexistspath-callback. This is why I opted for `existsSync` as awaiting the promise is effectively blocking too, so I think it's rather similar.

The documentation suggests using `fs.stat()` or `fs.access()` for checking the existence asynchronously, but I found that kinda clunky: you can see an example of how that looks in `configure.ts` (https://github.com/storybookjs/storybook/pull/29126/files#diff-27fc97f1960770b2c6cfeeb0f0c1f852df6d4d3a190a3ea293c7c864e97dddab).

If that approach is preferred by the maintainers nonetheless, I'm not opposed to changing it! 😄

Comment (ziebam) [reply code/core/src/cli/helpers.ts:202]: Also, the documentation raises a valid concern related to existence checks:

"Using fs.exists() to check for the existence of a file before calling fs.open(), fs.readFile(), or fs.writeFile() is not recommended. Doing so introduces a race condition, since other processes may change the file's state between the two calls. Instead, user code should open/read/write the file directly and handle the error raised if the file does not exist."

I wonder if I should rewrite such usages into the suggested approach to avoid race conditions. Thoughts?

File: code/core/scripts/helpers/dependencies.ts
Comment (ziebam) [reply code/core/scripts/helpers/dependencies.ts]: Only changed the usages that were introduced in this PR as to reduce the scope and not pollute the diff.

File: code/builders/builder-vite/src/index.ts
Comment (ziebam) [reply code/builders/builder-vite/src/index.ts:88]: It was added in 16.7.0 and the API seems to have stayed rather consistent since then, so maybe it's fine to use it. The final call is up to the maintainers though. 🙂

**Comment by ziebam** [comment]
The failures seem unrelated to my PR, although I'm not sure since there's not much output. Are they expected, @JReinhold?

File: code/builders/builder-vite/src/index.ts
Comment (JReinhold) [reply code/builders/builder-vite/src/index.ts:88]: Thanks for the heads up. 🙏 It looks like it was experimental then finally stable based on that it would be extremely weird for them to remove/modify the "experimental" API in v18/v20, so I'm fine with keeping it.

Let's remove the comment(s).

File: code/core/src/common/utils/get-storybook-info.ts
Comment (ziebam) [reply code/core/src/common/utils/get-storybook-info.ts:94]: False positive as `existsSync` doesn't throw.

File: code/core/src/cli/helpers.ts
Comment (JReinhold) [reply code/core/src/cli/helpers.ts:200]: I'm not too concerned with that, let's keep it as close as possible to the existing flow. in practice these files here aren't being modified simultaneously by other processes.

File: code/core/src/core-server/build-static.ts
Comment (JReinhold) [inline code/core/src/core-server/build-static.ts]: what happens if the dir doesn't exist when calling `readdir()`?

**Review by JReinhold — COMMENTED** [review]
Fantastic job!

Just to make sure, have you ensured all of these APIs are available in Node 18, which we currently support?

File: code/core/src/core-server/build-static.ts
Comment (JReinhold) [reply code/core/src/core-server/build-static.ts]: ah, think you caught the same while I was mid-review.

Comment (ziebam) [reply code/core/src/core-server/build-static.ts]: I've got your CI to thank as it's what caught it. 😄

File: code/core/scripts/helpers/generateTypesMapperFiles.ts
Comment (ziebam) [reply code/core/scripts/helpers/generateTypesMapperFiles.ts:36]: I slightly changed the way I implemented this (as when the parent directory didn't exist, `openFile` was throwing an error). Is it more clear now or should I still comment it where applicable?

Comment (JReinhold) [reply code/core/scripts/helpers/generateTypesMapperFiles.ts:36]: Looks fine now

**Review by JReinhold — APPROVED** [review]

**Comment by ziebam** [comment]
> Just to make sure, have you ensured all of these APIs are available in Node 18, which we currently support?

Yup, everything I've used was introduced before Node 18. The only concern could be `fsPromises.cp` which is marked as experimental, but that's been resolved in another PR comment.

**Comment by JReinhold** [comment]
Yes it looks flaky, don't worry about it.

**Pull request merged**
